volatility/framework/symbols/windows/extensions/__init__.py

```diff
@@ -872,3 +874,304 @@ def get_wait_reason(self) -> str:
+            starting_sector = subsection.StartingSector
+            subsection_offset = starting_sector * 0x200
+
+            # Similar to the check in is_valid(), make sure the SubsectionBase is not page aligned.
+            #if subsection.SubsectionBase & self.PAGE_MASK == 0:
+            #    break
+
+            ptecount = 0
+            while ptecount < subsection.PtesInSubsection:
+                pte_offset = subsection.SubsectionBase + (mmpte_size * ptecount)
+                file_offset = subsection_offset + ptecount * 0x1000
+
+                try:
+                    mmpte = self.get_pte(pte_offset)
+                except exceptions.InvalidAddressException:
+                    ptecount += 1
+                    continue
+
+                # First we check if the entry is valid. If so, then we get the physical offset.
+                # The valid entries are actually handled by the hardware.
+                if mmpte.u.Hard.Valid == 1:
+                    physoffset = mmpte.u.Hard.PageFrameNumber << 12
+                    yield physoffset, file_offset, self.PAGE_SIZE
+
+                elif mmpte.u.Soft.Prototype == 1:
+                    if not is_64bit and not is_pae:
+                        subsection_offset = ((mmpte.u.Subsect.SubsectionAddressHigh << 7) | (
+                                mmpte.u.Subsect.SubsectionAddressLow << 3))
+
+                # If the entry is not a valid physical address then see if it is in transition.
+                elif mmpte.u.Trans.Transition == 1:
+                    physoffset = mmpte.u.Trans.PageFrameNumber << 12
+                    yield physoffset, file_offset, self.PAGE_SIZE
+
+                # Go to the next PTE entry
+                ptecount += 1
+
+            # Go to the next Subsection in the single-linked list
+            subsection = subsection.NextSubsection
+
+class VACB(objects.StructType):
+    """A class for _VACB structures"""
+
+    FILEOFFSET_MASK = 0xFFFFFFFFFFFF0000
+
+    def get_file_offset(self) -> int:
+        # The FileOffset member of VACB is used to denote the offset within the file where the
+        # view begins. Since all views are 256 KB in size, the bottom 16 bits are used to
+        # store the number of references to the view.
+        return self.Overlay.FileOffset.QuadPart & self.FILEOFFSET_MASK
+
+class SHARED_CACHE_MAP(objects.StructType):
+    """A class for _SHARED_CACHE_MAP structures"""
+
+    VACB_BLOCK = 0x40000
+    VACB_OFFSET_SHIFT = 18
+    VACB_LEVEL_SHIFT = 7
+    VACB_SIZE_OF_FIRST_LEVEL = 1 << (VACB_OFFSET_SHIFT + VACB_LEVEL_SHIFT)
+    VACB_ARRAY = 0x80
+
+    def is_valid(self) -> bool:
+        """Determine if the object is valid."""
+
+        if self.FileSize.QuadPart <= 0 or self.ValidDataLength.QuadPart <= 0:
+            return False
+
+        if self.SectionSize.QuadPart < 0 or ((self.FileSize.QuadPart < self.ValidDataLength.QuadPart) and (
+                   self.ValidDataLength.QuadPart != 0x7fffffffffffffff)):
```

Comment:
Similar sentiment. but I won't make you jump through hoops if it's troublesome at all to calculate something...